tree: extract new packages to break dependency of execgen on tree

yuzefovich opened the pull request:
**execgen: move BinaryOverloadHelper into colexecbase**

This is so that it's easier to break the dependency on `tree` from
`execgen`. Additionally, this commit makes some of the comments
template-time only.

Release note: None

**treecmp: extract ComparisonOperator into a separate package**

This commit extracts things that `execgen` depends on (like some
constants for the comparison operators as well as some other stuff) into
a new package `treecmp`. The goal is to break the dependency of
`execgen` on `tree`.

It was mostly a simple move of some code from `tree/expr.go` into
`treecmp/comparison_operator.go`. I had to export a few things. Also,
I had to make `ComparisonOperatorSymbol.Inverse` take the receiver as an
argument and left the function in `tree` (in order to not mess with
`FoldComparisonExpr`).

Release note: None

**treebin: extract BinaryOperator into a separate package**

This commit extracts things that `execgen` depends on (like some
constants for the binary operators as well as some other stuff) into
a new package `treebin`. The goal is to break the dependency of
`execgen` on `tree`.

It was a simple move of some code from `tree/expr.go` into
`treebin/binary_operator.go`. I only had to export `IsPadded` function.

Release note: None

**treewindow: extract some window constants into a new package out of tree**

This commit extracts some constants about the window functions into
a new package which allows us to finally break the dependency of
`execgen` on `tree`. Now the absence of such dependency is enforced with
`VerifyNoImports` in `colexec` package.

The only notable change that I had to make is making
`WindowFrameExclusion` not implement `tree.NodeFormatter` interface;
instead, the callsites that wanted to call `Format` on that type now
call `String` which provides the same behavior.

Release note: None

yuzefovich committed:
execgen: move BinaryOverloadHelper into colexecbase

This is so that it's easier to break the dependency on `tree` from
`execgen`. Additionally, this commit makes some of the comments
template-time only.

Release note: None

yuzefovich committed:
treecmp: extract ComparisonOperator into a separate package

This commit extracts things that `execgen` depends on (like some
constants for the comparison operators as well as some other stuff) into
a new package `treecmp`. The goal is to break the dependency of
`execgen` on `tree`.

It was mostly a simple move of some code from `tree/expr.go` into
`treecmp/comparison_operator.go`. I had to export a few things. Also,
I had to make `ComparisonOperatorSymbol.Inverse` take the receiver as an
argument and left the function in `tree` (in order to not mess with
`FoldComparisonExpr`).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76354)
<!-- Reviewable:end -->

yuzefovich committed:
treebin: extract BinaryOperator into a separate package

This commit extracts things that `execgen` depends on (like some
constants for the binary operators as well as some other stuff) into
a new package `treebin`. The goal is to break the dependency of
`execgen` on `tree`.

It was a simple move of some code from `tree/expr.go` into
`treebin/binary_operator.go`. I only had to export `IsPadded` function.

Release note: None

yuzefovich committed:
treewindow: extract some window constants into a new package out of tree

This commit extracts some constants about the window functions into
a new package which allows us to finally break the dependency of
`execgen` on `tree`. Now the absence of such dependency is enforced with
`VerifyNoImports` in `colexec` package.

The only notable change that I had to make is making
`WindowFrameExclusion` not implement `tree.NodeFormatter` interface;
instead, the callsites that wanted to call `Format` on that type now
call `String` which provides the same behavior.

Release note: None

ajwerner reviewed (approved):
My one nit is that it might be nice to put some package-level comments on the new packages. Otherwise, seems mechanical and <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 🚀

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 12 files at r1, 41 of 75 files at r2, 38 of 50 files at r3, 35 of 35 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76354#-:-MvZeq1m4PW1hgTpXBKe:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Added quick comments for all new packages (although I couldn't come up with some very intelligent to say there).

TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76354#-:-MvZrtlbDUwQV582jVfi:bqs4v22)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->
